tracing/trace_string.hh

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (C) 2021 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include <vector>
+#include <seastar/core/sstring.hh>
+#include "seastarx.hh"
+
+namespace tracing {
+
+enum class string_type : uint8_t
+{
+    none = 0,
+    prepare_stmt,
+    execute_stmt,
+    execute_prepared,
+    execute_batch,
+    alternator,
+    last
+};
+
+inline std::array<const char *, 1+static_cast<size_t>(string_type::last)> trace_string_data = {
+    "None",
+    "Preparing CQL3 query",
+    "Execute CQL3 query",
+    "Execute CQL3 prepared query",
+    "Execute batch of CQL3 queries",
+    "Alternator",
+    "Last"
+};
```

Comment:
I would love to change it to `tracing::begin(const char *)`!

The only problem is that I did not find a way to guarantee that the string that is incoming to the call has static storage duration. If it is fine, I will be happy to remove enum and replace it with const char *. My first implementation used exactly this simple variant.

Reply:
As I mentioned above, enum is the only way to guarantee the robustness of this API and it's a totally fine API. 
I really don't understand why you, @nyh @sitano , assume that it's a less preferred way than using strings.

There is a finite amount of different types of requests that can be traced - this asks for an enum in the corresponding API. 

In addition the benefits of using enum are not limited to robustness - it adds to readability too. E.g. when you look at that enum definition - you can see what kind of requests you can trace right now.

And back to your suggestion about putting a comment, @nyh: an API that requires a user to read a comment has a very high probability of getting misused. One of the reasons why we have types is avoid such problems at the compile time.

If we stretch your idea to extreme we may say that we can use void* for every parameter and put a lot of comments near functions explaining what kind of parameters and what their life time should be.

Reply:
@vladzcloudius I wouldn't call the "enum" thing a "totally fine API". What I don't like about it is that instead of tracing being a utility used by many components of the system (e.g., CQL, Alternator, Redis, etc.), it becomes _mixed_ with those components - tracing.hh suddenly has constants for Alternator, for several different type of CQL requests, etc. I'm not worried like Avi about recompilation, but I have a personal dislike for spreading related code over multiple files. We have an alternator/ directory, and Alternator-specific stuff should be sprinkled all over the system.

I want to stress again that C++ is **not** a safe language in the sense that Rust is. A user may, as you said, take an `std::string()` and call `c_str()` on it to get a `const char*` which has an unknown lifetime. But a sane user won't, and a good reviewer won't accept such code.

Again, there are other solutions ot this problem besides `const char*`. One similar to the original code is to have an `std::string` - which requires copying but not `format()` calls so probably has relatively small cost. Another solution is to have a lw_shared_ptr<const char*> which is very safe (doesn't have the problem you are worried about) and also quite efficient (the extra costs are extra per-cpu copies of the same short string, and integer increment/decrement on every call).

In my opinion, the "enum" solution is the clunkiest solution, and I like it the least.

Reply:
how about now?

API

```
void begin(const char* name, gms::inet_address client) {}
```

usage

```
tracing::begin(trace_state, "alternator", client_state.get_client_address());
```

also there is a comment that the func expects static storage duration for its first argument const char *.

Reply:
> What I don't like about it is that instead of tracing being a utility used by many components of the system (e.g., CQL, Alternator, Redis, etc.), it becomes mixed with those components - tracing.hh suddenly has constants for Alternator, for several different type of CQL requests, etc.

@nyh I'm afraid that you are probably confusing the tracing::begin() with the tracing::trace(..).
The former denotes the beginning of the time counting and didn't even need to receive any string (more on this later). The later is similar to `printf` and can accept any string you want.

We are discussing the former here while I think what you have in mind is the later.

So, why tracing::begin() even accepts any string today? 
The string that it receives is a well defined and a very static string that is meant to separate queries of one kind from another: e.g. "Preparing CQL3 query", "Execute CQL3 query" or "Execute CQL3 prepared query". This string is going to be written into the 'request' column of the system_traces.sessions table.

By no means this string is meant to have any dynamic info. - Dynamic info (like `op`) is supposed to be stored into the `parameters` column.

So, for the Alternator code the string that is supposed to be passed to `tracing::begin()` should be something like: "Executing Alternator query".

As you may see, the parameter `tracing::begin()` receives denotes traced queries classes supported by the Tracing framework. And therefore defining those Traced queries classes in the Tracing is not just "not mixing" the Tracing code with corresponding components - it's the write way to do so because those classes definition belongs to the Tracing code.

And since the amount of those classes is fixed and well defined - using enum for denoting those classes is an obvious choice.

I hope it makes more sense now.

Reply:
> > What I don't like about it is that instead of tracing being a utility used by many components of the system (e.g., CQL, Alternator, Redis, etc.), it becomes mixed with those components - tracing.hh suddenly has constants for Alternator, for several different type of CQL requests, etc.
> 
> @nyh I'm afraid that you are probably confusing the tracing::begin() with the tracing::trace(..).

No, I wasn't confusing those. My point wasn't about how and which function it's used but about source-code organization. I want, as much as possible, all the Alternator-specific code to be in the alternator/ subdirectory. Having alternator-specific code in trace.hh is ugly, annoying, and I claim - unnecessary.

> So, why tracing::begin() even accepts any string today?
> The string that it receives is a well defined and a very static string that is meant to separate queries of one kind from another: e.g. "Preparing CQL3 query", "Execute CQL3 query" or "Execute CQL3 prepared query". This string is going to be written into the 'request' column of the system_traces.sessions table.
> 
> By no means this string is meant to have any dynamic info. - Dynamic info (like `op`) is supposed to be stored into the `parameters` column.

Ok, then let's fix that. And @sitano actually did this and I didn't object to that (because "op" is indeed copied into another parameter as well). What I objected to was having to pick the "static string" from some central file. I think this is unnecessary.

> So, for the Alternator code the string that is supposed to be passed to `tracing::begin()` should be something like: "Executing Alternator query".

That's exactly what I suggested - a string constant, like "Executing Alternator Query" (or whatever) stored in the Alternator code. @sitano rightly said that if this is a "char *", he has no way to enforce this this string has static lifetime. I suggested a few options, but I also think that char * is fine - nobody will pass s.c_str() for some temporary s without reading the documentation first.

> As you may see, the parameter `tracing::begin()` receives denotes traced queries classes supported by the Tracing framework. And therefore defining those Traced queries classes in the Tracing is not just "not mixing" the Tracing code with corresponding components - it's the right way to do so because those classes definition belongs to the Tracing code.

If Alternator wants to use the math library's sin() function, do we need to edit the math library to register Alternator as a registered user? Of course not. Similarly, we shouldn't need to edit the tracing library to use it in Alternator, Redis, or whatever future APIs we want to add.

> And since the amount of those classes is fixed and well defined - using enum for denoting those classes is an obvious choice.

It is a choice, and I agree it's not a complete disaster (it's not like we add those APIs every day), but I still think it's not the right choice.

Reply:
> My point wasn't about how and which function it's used but about source-code organization. I want, as much as possible, all the Alternator-specific code to be in the alternator/ subdirectory. 

And I definitely agree with you on that.

> Ok, then let's fix that. And @sitano actually did this and I didn't object to that (because "op" is indeed copied into another parameter as well). What I objected to was having to pick the "static string" from some central file. I think this is unnecessary.

You are not going to pick a string - but use a corresponding enum value.
All Alternator (Redis, or other user) are supposed to know is that they set the proper "type" of the request in question - Tracing will use the proper string (or whatever else) to tag the corresponding trace. 
If a new tag needs to be added - it needs to be added. Alternatively you are welcome to use any of the existing tags.

> If Alternator wants to use the math library's sin() function, do we need to edit the math library to register Alternator as a registered user? 

Let me give you a different example - if Alternator needs another RPC verb it would require a change in the RPC library. Alternatively you can use existing verbs. This is what we have here.

> It is a choice, and I agree it's not a complete disaster (it's not like we add those APIs every day), but I still think it's not the right choice.

I hope my explanations above have changed your opinion on that, @nyh